File: mobile/lib/services/clipboard/clipboard_service.dart
// ignore_for_file: dangling_library_doc_comments, unnecessary_library_name

/// Clipboard service for Synk mobile.
///
/// Provides a platform-agnostic interface to the system clipboard with
/// change detection and deduplication of repeated content.

import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

/// Service for interacting with the system clipboard.
///
/// Tracks the last clipboard value to detect changes and prevent
/// re-syncing the same content repeatedly.
class ClipboardService {
  String? _lastReadValue;
  final _changeController = StreamController<String>.broadcast();
  Timer? _pollingTimer;

  /// Stream of clipboard changes (new content only).
  Stream<String> get clipboardChangeStream => _changeController.stream;

  /// Current clipboard value (last read).
  String? get lastReadValue => _lastReadValue;

  /// Reads the current clipboard content.
  ///
  /// Returns empty string if clipboard is empty or unavailable.
  Future<String> read() async {
    try {
      final value = await Clipboard.getData('text/plain');
      final text = value?.text ?? '';
      _lastReadValue = text;
      return text;
    } catch (e) {
      debugPrint('Failed to read clipboard: $e');
      return '';
    }
  }

  /// Writes content to the clipboard.
  ///
  /// [text] - The text to write.
  /// [markAsLast] - Whether to update _lastReadValue (set false for incoming sync to prevent loop)
  Future<bool> write(String text, {bool markAsLast = true}) async {
    try {
      await Clipboard.setData(ClipboardData(text: text));
      if (markAsLast) {
        _lastReadValue = text;
      }
      return true;
    } catch (e) {
      debugPrint('Failed to write clipboard: $e');
      return false;
    }
  }

  /// Starts polling for clipboard changes.
  ///
  /// [intervalMs] - Polling interval in milliseconds.
  void startPolling({int intervalMs = 1000}) {
    stopPolling();
    _pollingTimer = Timer.periodic(
      Duration(milliseconds: intervalMs),
      (_) async {
        final current = await read();
        if (current.isNotEmpty && current != _lastReadValue) {
          _lastReadValue = current;
          _changeController.add(current);
        }
      },
    );
  }

  /// Stops polling for clipboard changes.
  void stopPolling() {
    _pollingTimer?.cancel();
    _pollingTimer = null;
  }

  /// Disposes resources.
  void dispose() {
    stopPolling();
    _changeController.close();
  }
}

File: desktop/src/main/services/clipboard-service.ts
/**
 * Clipboard service for Synk desktop.
 *
 * Provides cross-platform clipboard access with change detection.
 * Uses Electron's clipboard API for native clipboard operations.
 */

import { EventEmitter } from 'events';
import { clipboard } from 'electron';

/** Events emitted by the clipboard service. */
export interface ClipboardServiceEvents {
  change: [text: string];
  error: [error: Error];
}

/**
 * Clipboard service for desktop.
 *
 * Monitors the system clipboard for changes and provides read/write operations.
 */
export class ClipboardService extends EventEmitter<ClipboardServiceEvents> {
  private pollTimer: NodeJS.Timeout | null = null;
  private lastContent = '';
  private readonly pollIntervalMs: number;
  private disposed = false;

  /**
   * Creates a clipboard service.
   *
   * @param pollIntervalMs - Polling interval in milliseconds (default: 1000)
   */
  constructor(pollIntervalMs: number = 1000) {
    super();
    this.pollIntervalMs = pollIntervalMs;
    this.lastContent = this.read() || '';
  }

  /** Last read clipboard value (for loop prevention). */
  get lastReadValue(): string {
    return this.lastContent;
  }

  /**
   * Reads the current clipboard text content.
   *
   * @returns The clipboard text, or empty string if not text
   */
  read(): string {
    try {
      const text = clipboard.readText();
      this.lastContent = text;
      return text;
    } catch (error) {
      this.emit('error', error as Error);
      return '';
    }
  }

  /**
   * Writes text to the clipboard.
   *
   * @param text - Text to write
   * @param markAsLast - Whether to update lastContent to prevent loop detection
   */
  write(text: string, markAsLast: boolean = true): void {
    try {
      clipboard.writeText(text);
      if (markAsLast) {
        this.lastContent = text;
      }
    } catch (error) {
      this.emit('error', error as Error);
    }
  }

  /**
   * Checks if clipboard contains text.
   *
   * @returns true if clipboard has text content
   */
  hasText(): boolean {
    return clipboard.readText().length > 0;
  }

  /**
   * Clears the clipboard.
   */
  clear(): void {
    clipboard.clear();
    this.lastContent = '';
  }

  /**
   * Starts polling for clipboard changes.
   */
  startPolling(): void {
    if (this.disposed || this.pollTimer) return;

    this.pollTimer = setInterval(() => {
      this.checkForChanges();
    }, this.pollIntervalMs);
  }

  /**
   * Stops polling for clipboard changes.
   */
  stopPolling(): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
  }

  /**
   * Checks for clipboard changes and emits event if changed.
   */
  private checkForChanges(): void {
    if (this.disposed) return;

    const currentContent = this.read();
    if (currentContent && currentContent !== this.lastContent) {
      this.lastContent = currentContent;
      this.emit('change', currentContent);
    }
  }

  /**
   * Disposes all resources.
   */
  dispose(): void {
    this.disposed = true;
    this.stopPolling();
    this.removeAllListeners();
  }
}